### PR TITLE
I'm back and I've re-added the best feature in the game!

### PR DESCRIPTION
**Pull request by Giacom**
Hey guys, sorry about my short absence but I've returned and I'm here to make sure everyone gets their favourite feature back.

## About The Pull Request

Removed for no reason, telescience is back and it allows you to use simple trigonometry you help the warden with his overstocked armoury problem. Science starts with a powerful teleportation device that uses telecrystals to allow anyone, or anything, to be flinged through space and end up at a specific point, determined by some trigonometry maths.

## Why It's Good For The Game

This feature was highly requested to be reimplemented and what better person to add it back than the original creator himself! Now, I know you might be asking about how powerful of a feature it is but I assure it it will not be abused that much since only a nerd would create an excel spreadsheet to do the maths for you.

Besides without science everything would be considered magic so what better power to give science than the power of teleportation.

## Changelog
:cl:
add: Re-added telescience! Please do not abuse this or the Warden can valid kill you.
/:cl:

### CODE REVIEW

**Comment by ArcaneMusic** [comment]
Please god why

**Comment by trollbreeder** [comment]
please tell me this is an april fools joke

**Comment by Rohesie** [comment]
What the hell, it's really Giacom

File: code/modules/telesci/bscrystal.dm [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	name = "bluespace crystal"
+	desc = "A glowing bluespace crystal, not much is known about how they work. It looks very delicate."
+	icon = 'icons/obj/telescience.dmi'
+	icon_state = "bluespace_crystal"
+	w_class = 1
+	origin_tech = "bluespace=4;materials=3"
+	var/blink_range = 8 // The teleport range when crushed/thrown at someone.
+
+/obj/item/bluespace_crystal/New()
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/bscrystal.dm:13]:
```suggestion
/obj/item/bluespace_crystal/Initialize()
	. = ..()
```

File: code/modules/telesci/bscrystal.dm [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	pixel_x = rand(-5, 5)
+	pixel_y = rand(-5, 5)
+
+/obj/item/bluespace_crystal/attack_self(var/mob/user)
+	blink_mob(user)
+	user.drop_item()
+	user.visible_message("<span class='notice'>[user] crushes the [src]!</span>")
+	del(src)
```
Comment (Rohesie) [inline code/modules/telesci/bscrystal.dm:21]:
```suggestion
	qdel(src)
```

File: code/modules/telesci/bscrystal.dm [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+/obj/item/bluespace_crystal/proc/blink_mob(var/mob/living/L)
+	do_teleport(L, get_turf(L), blink_range, asoundin = 'sound/effects/phasein.ogg')
+
+/obj/item/bluespace_crystal/throw_impact(atom/hit_atom)
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/bscrystal.dm:27]:
```suggestion
	. = ..()
```

File: code/modules/telesci/bscrystal.dm [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	if(isliving(hit_atom))
+		blink_mob(hit_atom)
+	del(src)
```
Comment (Rohesie) [inline code/modules/telesci/bscrystal.dm:30]:
```suggestion
	qdel(src)
```

File: code/modules/telesci/bscrystal.dm [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+// Artifical bluespace crystal, doesn't give you much research.
+
+/obj/item/bluespace_crystal/artificial
+	name = "artificial bluespace crystal"
+	desc = "An artificially made bluespace crystal, it looks delicate."
+	origin_tech = "bluespace=2"
+	blink_range = 4 // Not as good as the organic stuff!
```
Comment (Rohesie) [inline code/modules/telesci/bscrystal.dm:38]:
```suggestion
	blink_range = 4 // Not as good as the organic stuff!

```

File: code/modules/telesci/gps.dm [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	icon_state = "gps-c"
+	w_class = 2.0
+	flags = FPRINT | TABLEPASS
+	slot_flags = SLOT_BELT
+	origin_tech = "programming=2;engineering=2"
+	var/gpstag = "COM0"
+	var/emped = 0
+
+/obj/item/device/gps/New()
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/gps.dm:15]:
```suggestion
/obj/item/device/gps/Initialize()
	.. = .()
```

File: code/modules/telesci/gps.dm [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	GPS_list.Add(src)
+	name = "global positioning system ([gpstag])"
+	overlays += "working"
+
+/obj/item/device/gps/Del()
+	GPS_list.Remove(src)
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/gps.dm:22]:
```suggestion
/obj/item/device/gps/Destroy()
	GPS_list.Remove(src)
	return ..()
```

File: code/modules/telesci/gps.dm [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+			name = "global positioning system ([gpstag])"
+			attack_self(usr)
+
+/obj/item/device/gps/science
+	icon_state = "gps-s"
+	gpstag = "SCI0"
+
+/obj/item/device/gps/engineering
+	icon_state = "gps-e"
+	gpstag = "ENG0"
```
Comment (Rohesie) [inline code/modules/telesci/gps.dm:72]:
```suggestion
	gpstag = "ENG0"

```

File: code/modules/telesci/telepad.dm [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		else if(stage == 1)
+			playsound(src, 'sound/items/Screwdriver.ogg', 50, 1)
+			user << "<span class = 'caution'> You screw in the telepad's tracking beacon.</span>"
+			stage = 0
+	if(istype(W, /obj/item/weapon/weldingtool) && stage == 1)
+		playsound(src, 'sound/items/Welder.ogg', 50, 1)
+		user << "<span class = 'caution'> You disassemble the telepad.</span>"
+		new /obj/item/stack/sheet/metal(get_turf(src))
+		new /obj/item/stack/sheet/glass(get_turf(src))
+		del(src)
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:46]:
```suggestion
		qdel(src)
```

File: code/modules/telesci/telepad.dm [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	icon_state = "beacon"
+	item_state = "signaler"
+	origin_tech = "bluespace=3"
+
+/obj/item/device/telepad_beacon/attack_self(mob/user as mob)
+	if(user)
+		user << "<span class = 'caution'> Locked In</span>"
+		new /obj/machinery/telepad_cargo(user.loc)
+		playsound(src, 'sound/effects/pop.ogg', 100, 1, 1)
+		del(src)
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:62]:
```suggestion
		qdel(src)
```

File: code/modules/telesci/telepad.dm [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	var/obj/machinery/pad = null
+	var/last_charge = 30
+	var/mode = 0
+	var/rand_x = 0
+	var/rand_y = 0
+	var/emagged = 0
+	var/teleporting = 0
+
+/obj/item/weapon/rcs/New()
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:86]:
```suggestion
/obj/item/weapon/rcs/Initialize()
	. = ..()
```

File: code/modules/telesci/telepad.dm [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+	processing_objects.Add(src)
+/obj/item/weapon/rcs/examine()
+	desc = "Use this to send crates and closets to cargo telepads. There are [rcharges] charges left."
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:90]:
```suggestion
	. = ..()
	. += "Use this to send crates and closets to cargo telepads. There are [rcharges] charges left."
```

File: code/modules/telesci/telepad.dm [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+	var/last_charge = 30
+	var/mode = 0
+	var/rand_x = 0
+	var/rand_y = 0
+	var/emagged = 0
+	var/teleporting = 0
+
+/obj/item/weapon/rcs/New()
+	..()
+	processing_objects.Add(src)
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:87]:
```suggestion
	START_PROCESSING(SSobj, src)
```

File: code/modules/telesci/telepad.dm [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+/obj/item/weapon/rcs/examine()
+	desc = "Use this to send crates and closets to cargo telepads. There are [rcharges] charges left."
+	..()
+
+/obj/item/weapon/rcs/Del()
+	processing_objects.Remove(src)
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:94]:
```suggestion
/obj/item/weapon/rcs/Destroy()
	STOP_PROCESSING(SSobj, src)
	return ..()
```

File: code/modules/telesci/telepad.dm [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+			user << "<span class = 'caution'> You calibrate the telepad locator.</span>"
+
+/obj/item/weapon/rcs/attackby(obj/item/W, mob/user)
+	if(istype(W,  /obj/item/weapon/card/emag) && emagged == 0)
+		emagged = 1
+		var/datum/effect/effect/system/spark_spread/s = new /datum/effect/effect/system/spark_spread
+		s.set_up(5, 1, src)
+		s.start()
+		user << "<span class = 'caution'> You emag the RCS. Click on it to toggle between modes.</span>"
+		return
```
Comment (Rohesie) [inline code/modules/telesci/telepad.dm:122]:
```suggestion
		return

```

File: code/modules/telesci/telesci_computer.dm [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+
+	// Based on the power used
+	var/teleport_cooldown = 0
+	var/list/power_options = list(5, 10, 20, 25, 30, 40, 50, 80, 100) // every index requires a bluespace crystal
+	var/teleporting = 0
+	var/starting_crystals = 3
+	var/list/crystals = list()
+
+/obj/machinery/computer/telescience/New()
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:29]:
```suggestion
/obj/machinery/computer/telescience/Initialize()
	. = ..()
```

**Comment by Giacom** [comment]
I will fix the compile errors later.

File: code/modules/telesci/telesci_computer.dm [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+	link_telepad()
+	recalibrate()
+
+/obj/machinery/computer/telescience/Del()
+	eject()
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:35]:
```suggestion
/obj/machinery/computer/telescience/Destroy()
	eject()
	return ..()
```

File: code/modules/telesci/telesci_computer.dm [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+
+/obj/machinery/computer/telescience/examine()
+	..()
+	usr << "There are [crystals.len ? crystals.len : "no"] bluespace crystals in the crystal slots."
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:39]:
```suggestion
/obj/machinery/computer/telescience/examine()
	. = ..()
	. += "There are [crystals.len ? crystals.len : "no"] bluespace crystals in the crystal slots."
```

File: code/modules/telesci/telesci_computer.dm [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+
+/obj/machinery/computer/telescience/initialize()
+	..()
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:42]:
```suggestion
/obj/machinery/computer/telescience/Initialize()
	. = ..()
```

File: code/modules/telesci/telesci_computer.dm [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+	if(istype(W, /obj/item/bluespace_crystal))
+		if(crystals.len >= power_options.len)
+			user << "<span class='warning'>There are not enough crystal slots.</span>"
+			return
+		user.drop_item()
+		crystals += W
+		W.loc = null
+		user.visible_message("<span class='notice'>[user] inserts [W] into \the [src]'s crystal slot.</span>")
+	else
+		..()
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:76]:
```suggestion
		return ..()
```

File: code/modules/telesci/telesci_computer.dm [hunk 21]
```diff
@@ -0,0 +1,309 @@
...
+
+/obj/machinery/computer/telescience/attack_ai(mob/user)
+	src.attack_hand(user)
+
+/obj/machinery/computer/telescience/attack_hand(mob/user)
+	if(..())
+		return
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:83]:
```suggestion
	. = ..()
	if(.)
		return
```

File: code/modules/telesci/telesci_computer.dm [hunk 22]
```diff
@@ -0,0 +1,309 @@
...
+		eject()
+		temp_msg = "NOTICE:<BR>Bluespace crystals ejected."
+
+	updateDialog()
+
+/obj/machinery/computer/telescience/proc/recalibrate()
+	teles_left = rand(30, 40)
+	//angle_off = rand(-25, 25)
+	power_off = rand(-4, 0)
+	rotation_off = rand(-10, 10)
```
Comment (Rohesie) [inline code/modules/telesci/telesci_computer.dm:309]:
```suggestion
	rotation_off = rand(-10, 10)

```

**Comment by Rohesie** [comment]
Might actually make it pass Travis, compile and work with these web-edits.

**Comment by tralezab** [comment]
thanks giacom

**Comment by Coastfront** [comment]
time to make an excel spreadsheet to do the math for me

File: code/modules/telesci/gps.dm [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+
+/obj/item/device/gps/Del()
+	GPS_list.Remove(src)
+	..()
+
+/obj/item/device/gps/emp_act(severity)
+	emped = 1
+	overlays -= "working"
+	overlays += "emp"
+	spawn(300)
```
Comment (Qustinnus) [inline code/modules/telesci/gps.dm:28]: bruh

File: code/modules/telesci/telepad.dm [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+	if(istype(W, /obj/item/weapon/wrench))
+		anchored = 0
+		playsound(src, 'sound/items/Ratchet.ogg', 50, 1)
+		if(anchored)
+			anchored = 0
+			user << "<span class = 'caution'> The [src] can now be moved.</span>"
+		else if(!anchored)
+			anchored = 1
+			user << "<span class = 'caution'> The [src] is now secured.</span>"
+	if(istype(W, /obj/item/weapon/screwdriver))
```
Comment (Qustinnus) [inline code/modules/telesci/telepad.dm:32]: use tool act

File: code/modules/telesci/telepad.dm [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+		if(stage == 0)
+			playsound(src, 'sound/items/Screwdriver.ogg', 50, 1)
+			user << "<span class = 'caution'> You unscrew the telepad's tracking beacon.</span>"
+			stage = 1
+		else if(stage == 1)
+			playsound(src, 'sound/items/Screwdriver.ogg', 50, 1)
+			user << "<span class = 'caution'> You screw in the telepad's tracking beacon.</span>"
+			stage = 0
+	if(istype(W, /obj/item/weapon/weldingtool) && stage == 1)
```
Comment (Qustinnus) [inline code/modules/telesci/telepad.dm:41]: use tool act

File: code/modules/telesci/telepad.dm [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+/obj/item/device/telepad_beacon
+	name = "telepad beacon"
+	desc = "Use to warp in a cargo telepad."
+	icon = 'icons/obj/radio.dmi'
+	icon_state = "beacon"
+	item_state = "signaler"
+	origin_tech = "bluespace=3"
+
+/obj/item/device/telepad_beacon/attack_self(mob/user as mob)
+	if(user)
```
Comment (Qustinnus) [inline code/modules/telesci/telepad.dm:58]: early return

File: code/modules/telesci/telepad.dm [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+	if(rcharges > 10)
+		rcharges = 10
+	if(last_charge == 0)
+		rcharges++
+		last_charge = 30
+	else
+		last_charge--
+
+/obj/item/weapon/rcs/attack_self(mob/user)
+	if(emagged)
```
Comment (Qustinnus) [inline code/modules/telesci/telepad.dm:105]: early return

File: code/modules/telesci/telesci_computer.dm [hunk 28]
```diff
@@ -0,0 +1,309 @@
...
+	power = power_options[1]
+
+/obj/machinery/computer/telescience/proc/link_telepad()
+	telepad = locate() in range(src, 7)
+
+/obj/machinery/computer/telescience/update_icon()
+	if(stat & BROKEN)
+		icon_state = "telescib"
+	else
+		if(stat & NOPOWER)
```
Comment (Qustinnus) [inline code/modules/telesci/telesci_computer.dm:55]: bruh just use else if wtf

File: code/modules/telesci/telesci_computer.dm [hunk 29]
```diff
@@ -0,0 +1,309 @@
...
+		else
+			icon_state = initial(icon_state)
+			stat &= ~NOPOWER
+
+/obj/machinery/computer/telescience/attack_paw(mob/user)
+	user << "You are too primitive to use this computer."
+	return
+
+/obj/machinery/computer/telescience/attackby(obj/item/W, mob/user)
+	if(istype(W, /obj/item/bluespace_crystal))
```
Comment (Qustinnus) [inline code/modules/telesci/telesci_computer.dm:67]: early return

File: code/modules/telesci/telesci_computer.dm [hunk 30]
```diff
@@ -0,0 +1,309 @@
...
+		flick("pad-beam", telepad)
+
+		if(spawn_time > 15) // 1.5 seconds
+			playsound(telepad.loc, 'sound/weapons/flash.ogg', 25, 1)
+			// Wait depending on the time the projectile took to get there
+			teleporting = 1
+			temp_msg = "Powering up bluespace crystals.<BR>Please wait."
+
+
+		spawn(round(proj_data.time) * 10) // in seconds
```
Comment (Qustinnus) [inline code/modules/telesci/telesci_computer.dm:175]: did you seriously just copy this archaic code from 2017 or something? dont use spawn and sleep

File: code/modules/telesci/telesci_computer.dm [hunk 31]
```diff
@@ -0,0 +1,309 @@
...
+
+			var/turf/source = target
+			var/turf/dest = get_turf(telepad)
+			if(sending)
+				source = dest
+				dest = target
+
+			flick("pad-beam", telepad)
+			playsound(telepad.loc, 'sound/weapons/emitter2.ogg', 25, 1)
+			for(var/atom/movable/ROI in source)
```
Comment (Qustinnus) [inline code/modules/telesci/telesci_computer.dm:211]: holy shit this entire proc needs some early returns

**Comment by anconfuzedrock** [comment]
this is thebest joke

**Comment by Rohesie** [comment]
This is an April's first PR, @Qustinnus 
My suggestions are to make the code compile and work so it can be TMed
Don't waste your time reviewing it for actual code standards

**Comment by Iamgoofball** [comment]
Yeaaaaah, welcome back giacom :)

**Comment by Iamgoofball** [comment]
>what better person to add it back than the original creator himself! 

Hey wait hold up

File: code/modules/telesci/gps.dm (hunk 6 above)
Comment (Dennok) [reply code/modules/telesci/gps.dm:15]:
```suggestion
/obj/item/device/gps/Initialize()
	. = ..()
```

**Comment by Giacom** [comment]
Original code please do not steal.

**Comment by Tharcoonvagh** [comment]
cool

**Comment by Giacom** [comment]
I must depart but I will be back again one day.

**Comment by Rohesie** [comment]
Sad!

**Comment by cacogen** [comment]
missed viacom

**Pull request closed without merging**